File: AlarmClock/app/src/main/java/com/example/alarmclock/footstepcount.java
package com.example.alarmclock;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

public class footstepcount extends AppCompatActivity implements SensorEventListener {
    SensorManager sensorManager;
    Sensor sensor;
    TextView stepTextView;

    boolean first = true;
    boolean up = false;
    float d0,d = 0f;
    int stepcount = 0;
//    フィルタリング係数 0<a<1
    float a = 0.90f;

   public footstepcount(){
       sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);
       sensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);

   }


    @Override
    protected void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.footstepcount);


        stepTextView = (TextView) findViewById(R.id.counter);
    }

    public void onSensorChanged(SensorEvent sensorEvent){
        float[] value = sensorEvent.values;
        float sum= (float)Math.sqrt(Math.pow(value[0],2) + Math.pow(value[1],2) + Math.pow(value[2],2));


        if (first){
            first = false;
            up = true;
            d0 = a * sum;
        }else{
            //ローパスフィルタリング 時系列の細かいデータを平滑化
            d =  a * sum + (1 - a) * d0;
            if (up && d < d0){
                up = false;
                stepcount++;




            }
            else if(!up&& d>d0){
                up = true;
                d0 = d;
            }
        }
        stepTextView.setText(stepcount + "歩");
    }
    public void onAccuracyChanged(Sensor sensor,int accuracy){

    }
    protected void onResume(){
        super.onResume();
    }
    protected void onPause(){
        super.onPause();
        sensorManager.unregisterListener(this);
    }
    public void clickStartButton(View view){
        sensorManager.registerListener(this,sensor,sensorManager.SENSOR_DELAY_GAME);
    }
    public void clickRestartButton(View view){
        stepTextView.setText("0歩");
        first = true;
        stepcount = 0;
    }
    public void clickStopButton(View view){
        this.onPause();
    }
}

File: AlarmClock/app/src/main/java/com/example/alarmclock/footstepcount.kt
package com.example.alarmclock

import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import java.text.SimpleDateFormat
import java.util.*

class footstepcount {
    class MainActivity : AppCompatActivity() , CoroutineScope, SensorEventListener {
        private lateinit var mManager: SensorManager
        private lateinit var mSensor: Sensor

        private var isJogging = false
        private var steps = 0
        private var startSteps = 0

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            setContentView(R.layout.activity_main)

            mManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager
            mSensor = mManager.getDefaultSensor(Sensor.TYPE_STEP_DETECTOR)

            val textView = findViewById<TextView>(R.id.textView)
            val button = findViewById<Button>(R.id.button)
            button.setOnClickListener {
                if (isJogging) {
                    finishJogging(it as Button, textView)
                } else {
                    startJogging(it as Button, textView)
                }
                isJogging = !isJogging
            }
        }

        private fun startJogging(button: Button, textView:TextView) {
            textView.text =""
            startSteps = this.steps
            button.text = resources.getText(R.string.finish)
        }

        private fun finishJogging(button: Button, textView:TextView) {
            button.text = resources.getText(R.string.now_sending)

            launch {
                val message = "${steps - startSteps}歩、歩きました！"
                val df = SimpleDateFormat("yyyy年MM月dd日HH持mm分", Locale.JAPAN)
                // ツイートする機能を入れる場所
                textView.text = message
                button.text = resources.getText(R.string.start)
            }
        }

        override fun onPause() {
            super.onPause()
            mManager.unregisterListener(this)
        }

        override fun onResume() {
            super.onResume()
            mManager.registerListener(this, mSensor, SensorManager.SENSOR_DELAY_NORMAL)
        }

        override fun onDestroy() {
            job.cancel()
            super.onDestroy()
        }

        override fun onSensorChanged(event: SensorEvent?) {
            steps++
        }

        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        }
    }







}